File: src/main/java/io/github/divyeshmangla/timetable/parser/ClassExtractor.java
package io.github.divyeshmangla.timetable.parser;

import io.github.divyeshmangla.timetable.model.ClassInfo;
import io.github.divyeshmangla.timetable.parser.reader.BlockClassReader;
import io.github.divyeshmangla.timetable.parser.reader.ClassReader;
import io.github.divyeshmangla.timetable.parser.reader.LargeClassReader;
import io.github.divyeshmangla.timetable.parser.reader.SingleClassReader;
import org.apache.poi.ss.usermodel.Cell;

import java.util.List;

/**
 * Orchestrates class extraction by trying each reader in order.
 */
public class ClassExtractor {

    private final List<ClassReader> readers = List.of(
            new SingleClassReader(),
            new LargeClassReader(),
            new BlockClassReader()
    );

    public ClassInfo extract(Cell cell) {
        for (ClassReader reader : readers) {
            if (reader.matches(cell)) {
                return reader.read(cell);
            }
        }
        return null;
    }
}

File: internal/parser/extractor/class_extractor.go
package extractor

import (
	"github.com/DivyeshMangla/tiet-timetable/internal/model"
	"github.com/DivyeshMangla/tiet-timetable/internal/parser/extractor/readers"
	"github.com/xuri/excelize/v2"
)

// ClassExtractor attempts to extract class information from a cell using
// multiple reader strategies. Readers are tried in order until one matches.
type ClassExtractor struct {
	readers []readers.Reader
}

func NewClassExtractor() *ClassExtractor {
	return &ClassExtractor{
		readers: []readers.Reader{
			&readers.SingleClassReader{},
			&readers.LargeClassReader{},
			&readers.BlockClassReader{},
			&readers.AlternatingLargeClassReader{},
		},
	}
}

func (ce *ClassExtractor) Extract(file *excelize.File, sheetName string, row, col int) *model.ClassInfo {
	for _, reader := range ce.readers {
		if matched, classInfo := reader.Read(file, sheetName, row, col); matched {
			return classInfo
		}
	}
	return nil
}
